File: ll_benchmark.py
# not verified, gpt generated.
import random
import time

LIST_SIZE = 100_000_000

class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def append(self, val):
        new_node = Node(val)
        if not self.head:
            self.head = self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node
        self.size += 1

    def insert(self, val, index):
        new_node = Node(val)
        if index == 0:
            new_node.next = self.head
            self.head = new_node
        else:
            curr = self.head
            for _ in range(index - 1):
                curr = curr.next
            new_node.next = curr.next
            curr.next = new_node
        self.size += 1

    def delete(self, index):
        if index == 0:
            self.head = self.head.next
        else:
            curr = self.head
            for _ in range(index - 1):
                curr = curr.next
            curr.next = curr.next.next
        self.size -= 1

    def __len__(self):
        return self.size

    def __iter__(self):
        curr = self.head
        while curr:
            yield curr.val
            curr = curr.next

def benchmark_lookup_by_value(lst):
    it = iter(lst)
    target1 = next(it)
    for _ in range(LIST_SIZE // 4):
        next(it)
    target2 = next(it)
    for _ in range(LIST_SIZE // 4):
        next(it)
    target3 = next(it)
    for _ in range(LIST_SIZE // 4):
        next(it)
    target4 = next(it)
    for _ in range(LIST_SIZE // 4):
        next(it)
    target5 = next(it)

    start = time.perf_counter_ns()
    [value for value in lst if value in (target1, target2, target3, target4, target5)]
    end = time.perf_counter_ns()

    print(f"Lookup By Value Time: {(end - start) // 1_000_000}ms")
    print(f"In nanosecs: {end - start}ns")

def benchmark_insert(lst):
    it = iter(lst)
    next(it)
    lst.insert(0, 0)
    for _ in range(LIST_SIZE // 4 - 1):
        next(it)
    lst.insert(LIST_SIZE // 4 + 1, 0)
    for _ in range(LIST_SIZE // 4):
        next(it)
    lst.insert(LIST_SIZE // 2 + 1, 0)
    for _ in range(LIST_SIZE // 4):
        next(it)
    lst.insert(3 * LIST_SIZE // 4 + 1, 0)
    for _ in range((LIST_SIZE // 4) - 1):
        next(it)
    lst.append(0)
    value_to_insert = 1

    start = time.perf_counter_ns()
    lst.insert(0, value_to_insert)
    end = time.perf_counter_ns()
    print(f"Insert (Beginning) Time: {end - start}ns")

    start = time.perf_counter_ns()
    lst.insert(LIST_SIZE // 4 + 1, value_to_insert)
    end = time.perf_counter_ns()
    print(f"Insert (q1) Time: {(end - start) // 1_000_000}ms")

    start = time.perf_counter_ns()
    lst.insert(LIST_SIZE // 2 + 1, value_to_insert)
    end = time.perf_counter_ns()
    print(f"Insert (Middle) Time: {(end - start) // 1_000_000}ms")

    start = time.perf_counter_ns()
    lst.insert(3 * LIST_SIZE // 4 + 1, value_to_insert)
    end = time.perf_counter_ns()
    print(f"Insert (q3) Time: {(end - start) // 1_000_000}ms")

    start = time.perf_counter_ns()
    lst.append(value_to_insert)
    end = time.perf_counter_ns()
    print(f"Insert (End) Time: {end - start}ns")

def benchmark_delete(lst):
    start = time.perf_counter_ns()
    lst.delete(0)
    end = time.perf_counter_ns()
    print(f"Delete (Beginning) Time: {end - start}ns")

    start = time.perf_counter_ns()
    lst.delete(lst.index(1))
    end = time.perf_counter_ns()
    print(f"Delete (q1) Time: {(end - start) // 1_000_000}ms")

    start = time.perf_counter_ns()
    lst.delete(lst.index(1))
    end = time.perf_counter_ns()
    print(f"Delete (Middle) Time: {(end - start) // 1_000_000}ms")

    start = time.perf_counter_ns()
    lst.delete(lst.index(1))
    end = time.perf_counter_ns()
    print(f"Delete (q3) Time: {(end - start) // 1_000_000}ms")

    start = time.perf_counter_ns()
    lst.delete(len(lst) - 1)
    end = time.perf_counter_ns()
    print(f"Delete (End) Time: {end - start}ns")

def main():
    lst = LinkedList()

    start = time.perf_counter_ns()
    for _ in range(LIST_SIZE):
        lst.append(random.randint(2, 2**31 - 1))
    end = time.perf_counter_ns()
    print(f"Pre-allocate Time: {(end - start) // 1_000_000}ms")

    run = 2
    while run > 0:
        benchmark_lookup_by_value(lst)
        benchmark_insert(lst)
        benchmark_delete(lst)
        run -= 1

    start = time.perf_counter_ns()
    insertions = 100
    step = len(lst) // insertions
    it = iter(lst)
    count = 0
    for val in lst:
        if count % step == 0:
            lst.insert(count, 0)
        count += 1
    for i, val in enumerate(lst):
        if val == 0:
            lst.insert(i + 1, 1)
    end = time.perf_counter_ns()
    print(f"Heavy Time: {(end - start) // 1_000_000}ms")

    start = time.perf_counter_ns()
    deletions = 100
    while deletions > 0:
        lst.delete(lst.index(1))
        deletions -= 1
    end = time.perf_counter_ns()
    print(f"Heavy Delete Time: {(end - start) // 1_000_000}ms")

    start = time.perf_counter_ns()
    sum = 0
    for val in lst:
        sum += val
    end = time.perf_counter_ns()
    print(f"{sum}")
    print(f"Heavy Lookup Time: {(end - start) // 1_000_000}ms")

if __name__ == "__main__":
    main()

File: ll_benchmark.java
import java.util.LinkedList;
import java.util.ListIterator;
import java.util.Random;

public class ll_benchmark {
    private static final int LIST_SIZE = 100_000_000;

    public static void benchmarkLookupByValue(LinkedList<Integer> lst) {
        ListIterator<Integer> it = lst.listIterator();
        int target1 = it.next();
        for (int i = 0; i < LIST_SIZE / 4 - 1; i++) {
            it.next();
        }
        int target2 = it.next();
        for (int i = 0; i < LIST_SIZE / 4 - 1; i++) {
            it.next();
        }
        int target3 = it.next();
        for (int i = 0; i < LIST_SIZE / 4 - 1; i++) {
            it.next();
        }
        int target4 = it.next();
        for (int i = 0; i < LIST_SIZE / 4 - 2; i++) {
            it.next();
        }
        int target5 = it.next();

        long start = System.nanoTime();
        lst.indexOf(target1);
        lst.indexOf(target2);
        lst.indexOf(target3);
        lst.indexOf(target4);
        lst.indexOf(target5);
        long end = System.nanoTime();

        System.out.println("Lookup By Value Time: " + (end - start) / 1_000_000 + "ms");
        System.out.println("In nanosecs: " + (end - start) + "ns");
    }

    public static void benchmarkInsert(LinkedList<Integer> lst) {
        ListIterator<Integer> it = lst.listIterator();
        int valueToInsert = 1;
        
        long start = System.nanoTime();
        lst.addFirst(valueToInsert);
        long end = System.nanoTime();
        System.out.println("Insert (Beginning) Time: " + (end - start) + "ns");

        start = System.nanoTime();
        it = lst.listIterator(LIST_SIZE / 4);
        it.next();
        it.add(valueToInsert);
        end = System.nanoTime();
        System.out.println("Insert (q1) Time: " + (end - start) / 1_000_000 + "ms");

        start = System.nanoTime();
        it = lst.listIterator(LIST_SIZE / 2);
        it.next();
        it.add(valueToInsert);
        end = System.nanoTime();
        System.out.println("Insert (Middle) Time: " + (end - start) / 1_000_000 + "ms");

        start = System.nanoTime();
        it = lst.listIterator(3 * LIST_SIZE / 4);
        it.next();
        it.add(valueToInsert);
        end = System.nanoTime();
        System.out.println("Insert (q3) Time: " + (end - start) / 1_000_000 + "ms");

        start = System.nanoTime();
        lst.addLast(valueToInsert);
        end = System.nanoTime();
        System.out.println("Insert (End) Time: " + (end - start) + "ns");
    }

    public static void benchmarkDelete(LinkedList<Integer> lst) {
        long start = System.nanoTime();
        lst.removeFirst();
        long end = System.nanoTime();
        System.out.println("Delete (Beginning) Time: " + (end - start) + "ns");

        start = System.nanoTime();
        int index = lst.indexOf(1);
        lst.remove(index);
        end = System.nanoTime();
        System.out.println("Delete (q1) Time: " + (end - start) / 1_000_000 + "ms");

        start = System.nanoTime();
        index = lst.indexOf(1);
        lst.remove(index);
        end = System.nanoTime();
        System.out.println("Delete (Middle) Time: " + (end - start) / 1_000_000 + "ms");

        start = System.nanoTime();
        index = lst.indexOf(1);
        lst.remove(index);
        end = System.nanoTime();
        System.out.println("Delete (q3) Time: " + (end - start) / 1_000_000 + "ms");

        start = System.nanoTime();
        lst.removeLast();
        end = System.nanoTime();
        System.out.println("Delete (End) Time: " + (end - start) + "ns");
    }

    public static void main(String[] args) {
        Random random = new Random();
        LinkedList<Integer> lst = new LinkedList<>();

        long start = System.nanoTime();
        for (int i = 0; i < LIST_SIZE; i++) {
            lst.add(random.nextInt(Integer.MAX_VALUE - 2) + 2);
        }
        long end = System.nanoTime();
        System.out.println("Pre-allocate Time: " + (end - start) / 1_000_000 + "ms");

        int run = 2;
        while (run > 0) {
            benchmarkLookupByValue(lst);
            benchmarkInsert(lst);
            benchmarkDelete(lst);
            run--;
        }

        start = System.nanoTime();
        int insertions = 100;
        int step = lst.size() / insertions;
        ListIterator<Integer> it = lst.listIterator();
        int count = 0;
        while (it.hasNext()) {
            it.next();
            if (count % step == 0) {
                it.set(0);
            }
            count++;
        }

        // Insert 1s after the 0s
        it = lst.listIterator();
        while (it.hasNext()) {
            int value = it.next();
            if (value == 0) {
                it.add(1);
            }
        }
        end = System.nanoTime();
        System.out.println("Heavy Time: " + (end - start) / 1_000_000 + "ms");

        start = System.nanoTime();
        int deletions = 100;
        while (deletions > 0) {
            int index = lst.indexOf(1);
            lst.remove(index);
            deletions--;
        }
        end = System.nanoTime();
        System.out.println("Heavy Delete Time: " + (end - start) / 1_000_000 + "ms");

        start = System.nanoTime();
        long sum = 0;
        for (Integer value : lst) {
            sum += value;
        }
        end = System.nanoTime();
        System.out.println(sum);
        System.out.println("Heavy Lookup Time: " + (end - start) / 1_000_000 + "ms");
    }
}

// Sequential 2 shot 100 mil
// Memory 4.5GB 
// CPU MAX~ 86%
// Pre-allocate Time: 29904ms
// Lookup By Value Time: 1552ms
// In nanosecs: 1552128500ns
// Insert (Beginning) Time: 52800ns
// Insert (q1) Time: 164ms
// Insert (Middle) Time: 326ms
// Insert (q3) Time: 143ms
// Insert (End) Time: 9100ns
// Delete (Beginning) Time: 5100ns
// Delete (q1) Time: 314ms
// Delete (Middle) Time: 637ms
// Delete (q3) Time: 604ms
// Delete (End) Time: 21200ns
// Lookup By Value Time: 1543ms
// In nanosecs: 1543066000ns
// Insert (Beginning) Time: 2200ns
// Insert (q1) Time: 164ms
// Insert (Middle) Time: 324ms
// Insert (q3) Time: 142ms
// Insert (End) Time: 1000ns
// Delete (Beginning) Time: 1200ns
// Delete (q1) Time: 316ms
// Delete (Middle) Time: 639ms
// Delete (q3) Time: 605ms
// Delete (End) Time: 1200ns
// Heavy Time: 1579ms
// Heavy Delete Time: 46175ms
// 107380387961924400
// Heavy Lookup Time: 775ms

// Sequential 2 shot 10 mil
// Pre-allocate Time: 2321ms
// Lookup By Value Time: 150ms
// In nanosecs: 150937100ns
// Insert (Beginning) Time: 47400ns
// Insert (q1) Time: 18ms
// Insert (Middle) Time: 35ms
// Insert (q3) Time: 14ms
// Insert (End) Time: 13600ns
// Delete (Beginning) Time: 6000ns
// Delete (q1) Time: 48ms
// Delete (Middle) Time: 64ms
// Delete (q3) Time: 55ms
// Delete (End) Time: 20900ns
// Lookup By Value Time: 160ms
// In nanosecs: 160693700ns
// Insert (Beginning) Time: 3700ns
// Insert (q1) Time: 19ms
// Insert (Middle) Time: 37ms
// Insert (q3) Time: 10ms
// Insert (End) Time: 3400ns
// Delete (Beginning) Time: 1100ns
// Delete (q1) Time: 37ms
// Delete (Middle) Time: 67ms
// Delete (q3) Time: 64ms
// Delete (End) Time: 1500ns
// Heavy Time: 153ms
// Heavy Delete Time: 4265ms
// 10739913489293855
// Heavy Lookup Time: 73ms